### DISCUSSION TITLE
Use Import of S Parameter file from SIWave

### DISCUSSION THREAD

**Discussion by kmahajan-cadfem** [ansys/pyedb#673] [Q&A]
Hi Everyone, 
we are very excited to use the PyEDB and try it out, for one of use cases we are looking for the following functionality from SIWave

![image](https://github.com/user-attachments/assets/d92e33ab-fa2b-455d-9b7b-054c25c314ed)

can maybe anyone comment, where can we find the above functionality in the API, unfortunately we could not find it using the documentation or the source code itself

**Comment by kmahajan-cadfem** [comment]
@RobPasMue and @SMoraisAnsys, could you please let me know if this is the right place to ask this questions or should I go to the [discuss.ansys.com](https://discuss.ansys.com/)

**Reply by RobPasMue** [reply]
I'd say it is the perfect place =)

**Comment by smedikon** [comment]
@svandenb-dev : FYI...
Please comment on this.

**Comment by svandenb-dev** [comment]
Hello @kmahajan-cadfem,
PyEDB supports Touchstone file assignment for RLC's that is straight forward. Here is a code snippet that should allow you to do this:

aedt_version = "2024.2"
source_edb = r"your_edb_path"
model_folder = r"your_s_param_model_folder"

edb = Edb(edbversion=aedt_version, edbpath=your_edb_path)
component_models = ["cmp_part1", "cmp_part2", "cmp_part3"]
for model in component_models:
    comp_model_folder = os.path.join(model_folder, model)
    if os.path.isdir(comp_model_folder):
        s_parameter_files = [f for f in os.listdir(comp_model_folder) if os.path.isfile(join(comp_model_folder, f)) and
                             pathlib.Path(f).suffix == ".s2p"]
        if s_parameter_files and model in edb.components.definitions:
            s_parameter_file = os.path.join(comp_model_folder, s_parameter_files[-1])
            if edb.components.definitions[model].assign_s_param_model(file_path=s_parameter_file,
                                                                      model_name=model):
                edb.logger.info(f"S-parameter model {s_parameter_file} assigned to component {model}")

I hope it helps feel free to contact us if any questions.

**Reply by SteffenSee92** [reply]
Hello @svandenb-dev ,

thanks for supporting us with the code snippet. As far as I understand the method you use in the script ("edb.components.definitions[model].**assign_s_param_model**(file_path=s_parameter_file,
model_name=model)"), this replaces the components by S-parameter blocks. When taking a look in SIwave, the corresponding parts are then no longer RLCs (with parasitics), but become "Discrete Devices". The corresponding functionality in the SIwave GUI would be the "Attach n-port S-parameter Model":
![image](https://github.com/user-attachments/assets/582676d2-cb35-4e05-971c-5aa514a1dda8)
The desired functionaliy in our case would be the left button "S-param Vendor Model Assignment":
![image](https://github.com/user-attachments/assets/fc890da6-1fe9-4ad2-8574-218ed9cc83fa)
This calculates parasitic RLC values from the S-parameters, when importing a Touchstone file via the button (1) and then assigns these calculated RLC values to the part (button 2):
![image](https://github.com/user-attachments/assets/1c5aaf6e-a832-4754-8e01-f40fc99d3001)
After the assignment, in SIwave the component is still treated as an RLC component and not as a discrete device, so it is still listed in the "Circuit Element Parameters" table.

Is there also a way to use the "Vendor Model Assignment" way from the SIwave GUI with scripting?

Thanks in advance!
Best regards
Steffen

**Reply by kmahajan-cadfem** [reply]
@svandenb-dev, would it be possible what  @SteffenSee92  has described above with help of the current status of `pyedb`, if not, is there any workaround for it?

**Comment by svandenb-dev** [comment]
Hi @kmahajan-cadfem @SteffenSee92 
I have implemented a new feature to load and use vendor libraires used by SIwave. This is currently published so please wait it is available.

edbapp = Edb()
comp_lib = edbapp.components.get_vendor_libraries()

This command is retrieving all LC located and CompLib/Locked folder used by siwave
example:

network = comp_lib.capacitors["AVX"]["AccuP01005"]["C005YJ0R1ABSTR"].s_parameters
it returns the scikit-rf Network object and we found a way to decode the binaries s parameters files. Once you get the Network object you can leverage the sckitkit-rf 
  [network (skrf.network) — scikit-rf Documentation](https://scikit-rf.readthedocs.io/en/latest/api/network.html#network-class)

and also export to *.s2p file
network.write_touchstone(os.path.join(edbapp.directory, "test_export.s2p"))

**Reply by kmahajan-cadfem** [reply]
Thanks, @SteffenSee92  will check this out later, give you a feedback. This would definitely help many customers as well 👍

**Reply by SteffenSee92** [reply]
Hi @svandenb-dev ,

thanks for sharing that example and the support so far.

Our desired goal with the scripting is to calculate the ESR and ESL values for capacitors from a given set of S-parameters s2p, the nominal C value and the information if it was a series or shunt measurement and in the end assign these calculated ESR and ESL values. As far as I can see from the SIwave GUI functionality, this calculation is implemented there and used, when one uses a model from the component library (predefined or own extra library). The import button within the "Vendor model assignement" menu simply adds the .rlc file with required information (shunt or series? Cnom?) for calculation of ESR and ESL from s2p. I summarized the desired workflow in the screenshot below.
Is the calculation implemented in SIwave also using the same calculation method that may be found in scikit-rf? Could you give us a hint what method to use from scikit for this Cnom, ESL and ESR approximation task?

Thanks, best regards
Steffen

![image](https://github.com/user-attachments/assets/ea04c994-c817-4504-92f0-985b76ee68ff)

**Reply by kmahajan-cadfem** [reply]
@svandenb-dev, any thoughts on the above, please let us know if it is at all possible or not, if not we can definitely create an issue for an enhancement and we can also communicate with the customer respectively.

**Reply by svandenb-dev** [reply]
@kmahajan-cadfem @SteffenSee92 
thanks for your feedback I will work on implemented ESL and ESL this would be valuable for lot of users.

**Reply by svandenb-dev** [reply]
@kmahajan-cadfem @SteffenSee92 
I have implemented properties to get ESL and ESR for capacitors only. The approach is first order, evaluating the self resonant frequency then evaluating the ESR and ESL at that frequency. My tests are showing good results but I couldn't check all values so please contact us if you find something. Please wait PyEDB v0.25.0 to be pubilshed.
edbapp = Edb()
comp_lib = edbapp.components.get_vendor_libraries()
network = comp_lib.capacitors["AVX"]["AccuP01005"]["C005YJ0R1ABSTR"].s_parameters
esr = comp_lib.capacitors["AVX"]["AccuP01005"]["C005YJ0R1ABSTR"].esr
esl = comp_lib.capacitors["AVX"]["AccuP01005"]["C005YJ0R1ABSTR"].esl

**Reply by SteffenSee92** [reply]
@svandenb-dev @kmahajan-cadfem 

Thanks for implementing that function.

I tested at first with the SIwave default library (complib in installation directory), and it seems to work in general. I checked a few entries (e.g., with the entry C005YJ0R1ABSTR shown in your example code) and compared to the values calculated by SIwave. The ESL values from both calculations match well, but with the ESR values there are differences in the range of a decade (with the C005YJ0R1ABSTR example, the script gives 1.76 ohms instead of 0.2 ohms for SIwave). It seems, SIwave uses a slightly different approximation approach here. Would it be possible to match both approaches?

I wanted to test then with an own library not located in the installation directory. This is a common approach also used by our customers simply due to missing admin rights for the installation directory...
I guess, the changed path could simply be handled by adapting the get_vendor_libraries() method, changing the path there. But the next problem I see, that this method only works with the original structure working with .txt and .sbin files. Is there any way to adapt this working with a combination of **.s2p and .rlc files**, which is the structure generated by SIwave when adding own elements to a capacitors library?

Thanks for your support!

**Reply by svandenb-dev** [reply]
@SteffenSee92 Sorry for the delay. Yes the approach is probably different. We will have to check the ESL ESR values evaluation. Like I mentioned a first order approximation by evaluating the self resonance.
